fission-web-server/library/Fission/Web/Server/Types.hs

```diff
@@ -397,15 +337,51 @@ instance IPFS.MonadLocalIPFS Server where
 
 instance IPFS.MonadRemoteIPFS Server where
   runRemote query = do
-    peerID       <- head <$> asks ipfsRemotePeers
-    IPFS.URL url <- asks ipfsURL
-    manager      <- asks httpManager
+    manager      <- asks ipfsHttpManager
+    urls         <- asks ipfsURLs
+    randomIndex  <- liftIO $ randomRIO (0, NonEmpty.length urls - 1)
+
+    let
+      IPFS.URL url  = urls NonEmpty.Partial.!! randomIndex -- Partial, but checked above
+      clientManager = mkClientEnv manager url
+
+    liftIO $ runClientM query clientManager
+
+instance MonadIPFSCluster Server where
+  runCluster query = do
+    clusterURLs     <- asks ipfsURLs
+    ipfsHttpManager <- asks ipfsHttpManager
+
+    logDebug @Text "🏘️  Running IPFS request across cluster (strict)"
+    forM clusterURLs \(IPFS.URL url) ->
+      liftIO . async $ runClientM query (mkClientEnv ipfsHttpManager url)
 
-    Peer.connectRetry peerID 25 >>= \case
-      Left  err -> logDebug $ "Unable to connect: " <> textDisplay err
-      Right _   -> return ()
+  streamCluster streamQuery = do
+    clusterURLs     <- asks ipfsURLs
+    ipfsHttpManager <- asks ipfsHttpManager
 
-    liftIO . runClientM query $ mkClientEnv manager url
+    logDebug @Text "🏘️  Running IPFS request across cluster (streaming)"
+    forM clusterURLs \(IPFS.URL url) -> do
+      resultChan <- liftIO newTChanIO
+      latestVar  <- atomically $ newTVar Nothing
+
+      asyncRef <- liftIO $ async do
+        Stream.withClientM streamQuery (mkClientEnv ipfsHttpManager url) \case
+          Left clientErr ->
+            return $ Left clientErr
+
+          Right ioSource -> do
+            runExceptT (Stream.runSourceT ioSource) >>= \case
+              Left errMsg ->
+                return . Left . ConnectionError . toException $ GenericError errMsg
+
+              Right vals ->
+                case List.lastMaybe vals of
+                  Nothing    -> return . Left . ConnectionError . toException $ NotFound @PinStatus
+                  Just final -> return $ Right final
+
+      asyncIdleTimeout (Seconds (Unity (120 :: Natural))) asyncRef latestVar
+      return (asyncRef, resultChan)
```

Comment:
This is fairly quick & dirty, but we're not doing any "real" stream processing at the moment, so pushing over TChans as a way to get live updates if needed is here for convenience.